// react-forms-ui/lib/fields/custom-field.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _label = require('../label/label');

var _label2 = _interopRequireDefault(_label);

var _messages = require('../messages/messages');

var _messages2 = _interopRequireDefault(_messages);

var _fieldMixin = require('./field-mixin');

var _fieldMixin2 = _interopRequireDefault(_fieldMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CustomField = _react2.default.createClass({
	displayName: 'CustomField',


	mixins: [_fieldMixin2.default],

	render: function render() {
		var _props = this.props,
		    id = _props.id,
		    label = _props.label,
		    required = _props.required,
		    readonly = _props.readonly,
		    children = _props.children;
		var classes = this.props.classes;

		classes = classes ? classes.split(',') : [];
		return _react2.default.createElement(
			'div',
			{ className: 'form-group' },
			_react2.default.createElement(
				_label2.default,
				{ htmlFor: id, className: classes[0], required: required ? 'required' : false },
				label
			),
			_react2.default.createElement(
				'div',
				{ className: classes[1] },
				readonly && _react2.default.createElement(
					'div',
					{ className: 'form-control-static' },
					children
				),
				!readonly && children
			),
			!readonly && _react2.default.createElement(_messages2.default, { ref: 'messages', id: id, fieldMessages: this._getFieldMessages(),
				showFeedback: this.state.showFeedback, className: classes[2] })
		);
	}
});

CustomField.contextTypes = {
	form: _react2.default.PropTypes.object
};

CustomField.propTypes = {
	id: _react2.default.PropTypes.string.isRequired,
	label: _react2.default.PropTypes.string.isRequired,
	required: _react2.default.PropTypes.bool,
	readonly: _react2.default.PropTypes.bool
};

exports.default = CustomField;
