// react-forms-ui/lib/messages/form-messages.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _messages = require('./messages');

var _messages2 = _interopRequireDefault(_messages);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FormMessages = _react2.default.createClass({
	displayName: 'FormMessages',
	getInitialState: function getInitialState() {
		return { showFeedback: 'none' };
	},
	componentWillMount: function componentWillMount() {
		var form = this.context.form;

		if (form) {
			form.registerField(this);
		}
	},
	render: function render() {
		var className = this.props.className;
		var showFeedback = this.state.showFeedback;
		var state = this.context.form.props.state;

		if (!state || !state.messages || !state.messages._form) {
			return _react2.default.createElement('div', null);
		}
		return _react2.default.createElement(
			'div',
			{ className: 'form-group' },
			_react2.default.createElement(_messages2.default, { id: '_form', fieldMessages: state.messages._form, showFeedback: showFeedback,
				className: className })
		);
	}
});

FormMessages.contextTypes = {
	form: _react2.default.PropTypes.object.isRequired
};

exports.default = FormMessages;
