// react-forms-ui/lib/controls/text-area-control.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _controlMixin = require('./control-mixin');

var _controlMixin2 = _interopRequireDefault(_controlMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

exports.default = _react2.default.createClass({
	displayName: 'text-area-control',


	mixins: [_controlMixin2.default],

	render: function render() {
		var _props = this.props,
		    id = _props.id,
		    readonly = _props.readonly,
		    placeholder = _props.placeholder,
		    label = _props.label,
		    value = _props.value,
		    _props$className = _props.className,
		    className = _props$className === undefined ? '' : _props$className,
		    formControl = _props.formControl,
		    _props$rows = _props.rows,
		    rows = _props$rows === undefined ? 3 : _props$rows,
		    onChange = _props.onChange,
		    onBlur = _props.onBlur,
		    children = _props.children,
		    otherProps = _objectWithoutProperties(_props, ['id', 'readonly', 'placeholder', 'label', 'value', 'className', 'formControl', 'rows', 'onChange', 'onBlur', 'children']);

		if (readonly) {
			var displayValue = value ? value.split('\n').map(function (line, index) {
				return _react2.default.createElement(
					'div',
					{ key: index },
					line
				);
			}) : value;
			if (formControl) {
				return _react2.default.createElement(
					'div',
					_extends({ className: className + ' form-control-static' }, otherProps),
					displayValue
				);
			}
			return _react2.default.createElement(
				'span',
				_extends({ className: className }, otherProps),
				displayValue
			);
		}
		return _react2.default.createElement(
			'span',
			null,
			_react2.default.createElement('textarea', _extends({ ref: 'input', id: id, name: id, className: className + ' form-control', rows: rows,
				placeholder: placeholder || label, value: value || '' }, otherProps, {
				onChange: this._onChange, onBlur: this._onBlur })),
			children
		);
	},
	focus: function focus() {
		var input = this.refs.input;

		if (input) {
			var el = _reactDom2.default.findDOMNode(input);
			el.focus();
		}
	},
	_onChange: function _onChange() {
		var onChange = this.props.onChange;

		if (onChange) {
			var value = this.refs.input.value;
			onChange(value);
		}
	},
	_onBlur: function _onBlur(event) {
		var onBlur = this.props.onBlur;

		if (onBlur) {
			onBlur(event);
		}
	}
});
